cdap-ui/app/features/adapters/controllers/detail-ctrl.js

```diff
@@ -15,12 +15,214 @@
+        title: 'Datasets',
+        template: '/assets/features/adapters/templates/tabs/datasets.html'
+      }
+    ];
+
+    $scope.activeTab = $scope.tabs[0];
+    $scope.selectTab = function(tab) {
+      $scope.activeTab = tab;
+    };
+
+
+    var params = {
+      namespace: $state.params.namespace,
+      appId: rAdapterDetail.name,
+      scope: $scope
+    };
+
+    if (AdapterDetail.programType === 'WORKFLOWS') {
+      angular.forEach(rAdapterDetail.programs, function (program) {
+        if (program.type === 'Workflow') {
+          params.workflowId = program.id;
+        }
+      });
+
+      myWorkFlowApi.pollStatus(params)
+        .$promise
+        .then(function (res) {
+          if (res.status === 'RUNNING') {
+            $scope.appStatus = res.status;
+          } else {
+            myWorkFlowApi.getScheduleStatus({
+              namespace: $state.params.namespace,
+              appId: rAdapterDetail.name,
+              scheduleId: 'etlWorkflow',
+              scope: $scope
+            })
+              .$promise
+              .then(function (schedule) {
+                $scope.appStatus = schedule.status;
+              });
+          }
+        });
+
+    } else {
+
+      angular.forEach(rAdapterDetail.programs, function (program) {
+        if (program.type === 'Workers') {
+          params.workerId = program.id;
+        }
+      });
+
+      myWorkersApi.pollStatus(params)
+        .$promise
+        .then(function (res) {
+          $scope.appStatus = res.status;
+        });
+    }
+
+
+    $scope.do = function (action) {
+      var scheduleParams = {
+        namespace: $state.params.namespace,
+        appId: rAdapterDetail.name,
+        scheduleId: 'etlWorkflow',
+        scope: $scope
+      };
+
+      switch (action) {
+        case 'Start':
+          $scope.appStatus = 'Starting';
+          if (AdapterDetail.programType === 'WORKFLOWS') {
+            myWorkFlowApi.scheduleResume(scheduleParams, {})
+              .$promise
+              .then(function () {
+                $scope.appStatus = 'SCHEDULED';
+              });
+          } else {
+            myWorkersApi.doAction(angular.extend(params, { action: 'start' }, {}))
+              .$promise
+              .then(function () {
+                $scope.appStatus = 'RUNNING';
+              }, function () {
+                $scope.appStatus = 'FAILED';
+              });
+          }
+
+          break;
+
+        case 'Stop':
+          $scope.appStatus = 'Stopping';
+          if (AdapterDetail.programType === 'WORKFLOWS') {
+            myWorkFlowApi.scheduleSuspend(scheduleParams, {})
+              .$promise
+              .then(function () {
+                $scope.appStatus = 'SUSPENDED';
+              });
+          } else {
+            myWorkersApi.doAction(angular.extend(params, { action: 'stop' }, {}))
+              .$promise
+              .then(function () {
+                $scope.appStatus = 'STOPPED';
+              });
```

Comment:
When backend returns 200 OK, it does mean the action is complete. It won't return anything if the backend is still stopping or resuming the schedule. And if something goes wrong, it will throw an error